e2e/scripts/common.func

```diff
@@ -0,0 +1,178 @@
+# generic functions
+
+fatal() {
+    test-error "$1"
+    exit 1
+}
+
+# test if singularity binary is present in $PATH
+if ! command -v singularity >/dev/null 2>&1; then
+    fatal "singularity executable not found in \$PATH"
+fi
+
+# execute/test commands interactively
+_uuid=`uuidgen`
+_script="${TESTDIR}/${_uuid}.script"
+_cmdline="${TESTDIR}/${_uuid}.cmdline"
+
+_check-script() {
+    if [ ! -f ${_script} ]; then
+        fatal "interactive function not used between start-interactive and exit-interactive calls"
+    fi
+}
+
+start-interactive() {
+    name=$1
+    shift
+    echo -n > ${_script}
+    echo "${name} ${_script} $@" > ${_cmdline}
+}
+
+interactive-send() {
+    _check-script
+    echo "send $1" >> ${_script}
+}
+
+interactive-expect() {
+    _check-script
+    echo "expect $1" >> ${_script}
+}
+
+interactive-sendline() {
+    _check-script
+    echo "sendline $1" >> ${_script}
+}
+
+interactive-ctrl-c() {
+    _check-script
+    interactive-send $'\x03'
+}
+
+interactive-ctrl-d() {
+    _check-script
+    interactive-send $'\x04'
+}
+
+interactive-eof() {
+    _check-script
+    interactive-expect "<EOF>"
+}
+
+exit-interactive() {
+    if [ $# != 1 ]; then
+        fatal "exit-interactive need an exit code argument"
+    fi
+    if [ ! -f ${_script} ]; then
+        fatal "no interactive script provided"
+    fi
+    if [ ! -f ${_cmdline} ]; then
+        fatal "no command line provided in start-interactive"
+    fi
+    expect-exit-interactive $1 `cat ${_cmdline}`
+    rm -f ${_script}
+    rm -f ${_cmdline}
+}
+
+# wrapper around default test builtin
+expect-failure() {
+    expect-exit 1 "$@"
+}
+
+expect-success() {
+    expect-exit 0 "$@"
+}
+
+stdout-contains() {
+    expect-search output "$@"
+}
+
+stderr-contains() {
+    expect-search error "$@"
+}
+
+test-match() {
+    expect-exit 0 $1 /bin/sh -c "[ \"$2\" = \"$3\" ]"
+}
+
+# manage test file/directory
+mktmpdir() {
+    create-tmpdir ${TESTDIR}
+}
+
+mktmpfile() {
+    create-tmpfile ${TESTDIR}
+}
+
+# test image
+get-test-image() {
```

Comment:
There are a lot of ways to get an image for testing, perhaps this should be something like `build-test-image-from-definition`?

Reply:
this is the equivalent of `e2e.EnsureImage` from e2e package, I could rename to `ensure-test-image` ?